stat/roc.go

```diff
@@ -127,3 +129,70 @@ func ROC(cutoffs, y []float64, classes []bool, weights []float64) (tpr, fpr, thr
 
 	return tpr, fpr, cutoffs
 }
+
+// TOC returns the Total Operating Characteristic for the classes provided
+// and the minimum and maximum bounds for the TOC.
+//
+// The input y values that correspond to classes and weights must be sorted
+// in ascending order. classes[i] is the class of value y[i] and weights[i]
+// is the weight of y[i]. SortWeightedLabeled can be used to sort classes
+// together with weights by the rank variable.
+//
+// The returned ntp values can be interpreted as the number of true positives
+// where values above the given rank are assigned class true for each given
+// rank from 1 to len(classes). The values of min and max provide the minimum
+// and maximum possible number of true values for the set of classes.
+//
+// If weights is nil, all weights are treated as 1. When weights are not nil,
+// the calculation of min and max allows for partial assignment of single data
+// points. If weights is not nil it must have the same length as classes,
+// otherwise TOC will panic.
+//
+// More details about TOC curves are available at
+// https://en.wikipedia.org/wiki/Total_operating_characteristic
+func TOC(classes []bool, weights []float64) (min, ntp, max []float64) {
+	if weights != nil && len(classes) != len(weights) {
+		panic("stat: slice length mismatch")
+	}
+	if len(classes) == 0 {
+		return nil, nil, nil
+	}
+
+	ntp = make([]float64, len(classes)+1)
+	min = make([]float64, len(ntp))
+	max = make([]float64, len(ntp))
+	if weights == nil {
+		for i := range ntp[1:] {
+			ntp[i+1] = ntp[i]
+			if !classes[i] {
```

Comment:
Can you elaborate?

Reply:
Number of true positives is the number of variables with class == True and classified as True, right? But if all classes == False, then ntp[i] == 0 by definition. But the code will be incrementing ntp[i + 1] in every step.

What am I not getting here?

Reply:
Yeah, this part does my head in, but it's a consequence of the decision to invert the parity of the ROC input. If we accept that the decision taken in #316 was correct, then this follows.

Reply:
But that's not really an answer, is it? What if the decision taken in #316 was wrong? Or maybe the consequence doesn't follow? Actually, looking at the discussion in #316 and the changes in #959, it looks to me that the change of parity you're talking about had to do with changing the interpretation of the `y` input to the `ROC` function, not of the `classes` input.

Maybe it all checks out and I'm wasting your time, but I would really feel safer approving this PR if I could see that the code works OK without referring to other PRs and other functions, just from the standard definition of TOC. As it is, even the updated docstring is clearly inconsistent with the code, unless we decided to redefine the meaning of "true positive"...

Reply:
Yes, I'm extremely uncomfortable with this (the part you edited out is absolutely correct).

The published definition of TOC (unfortunately an excel spread sheet is the implementation associated with the paper) is the inverse of the class parity here. However, it's important that the TOC AUC and ROC AUC agree since they are isomorphic. Under this transformation (flipping reality) the implementation here agrees with the paper.

The changes in #959 dealt with a few things, including the `classes` input. From memory much of that was driven by the stated desire to match the SciPy behaviour.

> As it is, even the updated docstring is clearly inconsistent with the code, unless we decided to redefine the meaning of "true positive"...

Yes. This makes me unhappy.

Reply:
So the discussion in #316 includes this statement by @btracey:

> ROC treats values below the cutoff as true (rather than false)

But this is different than what Wikipedia says: https://en.wikipedia.org/wiki/Receiver_operating_characteristic#Curves_in_ROC_space

> Given a threshold parameter T, the instance is classified as "positive" if X>T, and "negative" otherwise.

If we associate "positive" with "true" and "negative" with "false".

Reply:
And the documentation for sklearn says about the 2nd argument of `roc_auc_score`: https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html#sklearn.metrics.roc_auc_score

> In the binary case, it corresponds to an array of shape (n_samples,). Both probability estimates and non-thresholded decision values can be provided. The probability estimates correspond to the probability of the class with the greater label, i.e. estimator.classes_[1] and thus estimator.predict_proba(X, y)[:, 1]. The decision values corresponds to the output of estimator.decision_function(X, y). See more information in the User guide;

The user guide https://scikit-learn.org/stable/modules/model_evaluation.html#roc-metrics makes it clear that they treat the values above the threshold as True.

Looking at https://github.com/gonum/gonum/blob/master/stat/roc_example_test.go#L111, the current AUC function in gonum matches sklearn, hence it treats - like sklearn and Wikipedia - the values ABOVE the threshold as True.

Reply:
Yeah. the issue is the behaviour of SciKit learn (where [1 is positive](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_curve.html)). We see it do this

```
>>> from sklearn import metrics
>>> y = [0.1, 0.35, 0.4, 0.8]
>>> classes = [1, 0, 1, 0]
>>> metrics.roc_auc_score(classes, y)
0.25
```
and the (probably) [equivalent Gonum](https://github.com/gonum/gonum/blob/eae7c2a69e552715bd866b03ccd39d0c90504af6/stat/roc_example_test.go#L111-L127) code do this

```
y := []float64{0.1, 0.35, 0.4, 0.8}
classes := []bool{true, false, true, false}

tpr, fpr, _ := stat.ROC(nil, y, classes, nil)

// Compute Area Under Curve.
auc := integrate.Trapezoidal(fpr, tpr)
fmt.Printf("true  positive rate: %v\n", tpr)
fmt.Printf("false positive rate: %v\n", fpr)
fmt.Printf("auc: %v\n", auc)

// Output:
// true  positive rate: [0 0 0.5 0.5 1]
// false positive rate: [0 0.5 0.5 1 1]
// auc: 0.25
```

Reply:
In summary, I think the TOC function should:

1. treat values ABOVE the threshold as True
2. define `ntp[i]` as

```
ntp_i = sum_{j = i + 1}^{len(ntp) - 1} [ classes_j ] * weights_j, where [x] = 1 if x else 0.
ntp_{len(ntp) - 1} = 0
```

And then I think we will match AUC, sklearn and remain sane.

Reply:
> Yeah. the issue is the behaviour of SciKit learn (where [1 is positive](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_curve.html)). We see it do this
> 
> ```
> >>> from sklearn import metrics
> >>> y = [0.1, 0.35, 0.4, 0.8]
> >>> classes = [1, 0, 1, 0]
> >>> metrics.roc_auc_score(classes, y)
> 0.25
> ```
> 
> and the (probably) [equivalent Gonum](https://github.com/gonum/gonum/blob/eae7c2a69e552715bd866b03ccd39d0c90504af6/stat/roc_example_test.go#L111-L127) code do this
> 
> ```
> y := []float64{0.1, 0.35, 0.4, 0.8}
> classes := []bool{true, false, true, false}
> 
> tpr, fpr, _ := stat.ROC(nil, y, classes, nil)
> 
> // Compute Area Under Curve.
> auc := integrate.Trapezoidal(fpr, tpr)
> fmt.Printf("true  positive rate: %v\n", tpr)
> fmt.Printf("false positive rate: %v\n", fpr)
> fmt.Printf("auc: %v\n", auc)
> 
> // Output:
> // true  positive rate: [0 0 0.5 0.5 1]
> // false positive rate: [0 0.5 0.5 1 1]
> // auc: 0.25
> ```

We would probably gain total clarity if we also displayed the thresholds in the Gonum example.

https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_curve.html generates the following complete output:

```
>>> import numpy as np
>>> from sklearn import metrics
>>> y = np.array([1, 1, 2, 2])
>>> scores = np.array([0.1, 0.4, 0.35, 0.8])
>>> fpr, tpr, thresholds = metrics.roc_curve(y, scores, pos_label=2)
>>> fpr
array([0. , 0. , 0.5, 0.5, 1. ])
>>> tpr
array([0. , 0.5, 0.5, 1. , 1. ])
>>> thresholds
array([1.8 , 0.8 , 0.4 , 0.35, 0.1 ])
```

Which clearly shows that the TPR goes up as the threshold goes down.

Reply:
```
y := []float64{0.1, 0.4, 0.35, 0.8}
classes := []bool{false, false, true, true}
stat.SortWeightedLabeled(y, classes, nil)

tpr, fpr, thresh := stat.ROC(nil, y, classes, nil)
fmt.Printf("true  positive rate: %v\n", tpr)
fmt.Printf("false positive rate: %v\n", fpr)
fmt.Printf("threshold          : %v\n", thresh)
```

```
true  positive rate: [0 0.5 0.5 1 1]
false positive rate: [0 0 0.5 0.5 1]
threshold          : [+Inf 0.8 0.4 0.35 0.1]
```